### PR TITLE
Add `ParametersRef` to `Gateway.spec.infrastructure`

### PR DESCRIPTION
**Pull request by howardjohn**
**What type of PR is this?**

/kind gep

**What this PR does / why we need it**:

Follow up to https://github.com/kubernetes-sigs/gateway-api/pull/1868, which was down-scoped to only include labels and annotations to reduce friction on the initial GEP

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
A new `parametersRef` field has been added to `Gateway.spec.infrastructure`. This allows configuration of arbitrary implementation-specific parameters. This field previously existed only on `GatewayClass`. 
```

### CODE REVIEW

File: apis/v1/gateway_types.go
Comment (robscott) [inline apis/v1/gateway_types.go:667]: This allows for cross-namespace references, I'm not sure we want to allow for that. The [corresponding GEP](https://gateway-api.sigs.k8s.io/geps/gep-1867/) doesn't make it obvious if those are needed, do you mind updating the GEP to clarify the goals here?

File: apis/v1/gateway_types.go
Comment (robscott) [inline apis/v1/gateway_types.go:662]: I think we need to describe how merging works if/when this is configured at both layers.

**Review by robscott — COMMENTED** [review]
Thanks @howardjohn! Definitely a stretch to get into v1.1 given the timeline, but since this is just a part of a GEP that was already promoted to experimental, it seems like there might be a small chance it could fit in if we don't run into any significant issues.

File: apis/v1/gateway_types.go
Comment (howardjohn) [reply apis/v1/gateway_types.go]: I made a new `LocalParametersReference` to force it to be same namespace

File: apis/v1/gateway_types.go
Comment (howardjohn) [reply apis/v1/gateway_types.go:662]: Added

**Review by mikemorris — APPROVED** [review]
Would like to see this get in!

For context, [GEP-1867](https://gateway-api.sigs.k8s.io/geps/gep-1867/), which is currently in Experimental status, covers the background motivation and why GatewayClass `parametersRef` is not sufficient, and https://github.com/kubernetes-sigs/gateway-api/issues/2911 is an example of a specific issue that arose from the tension between how GatewayClass `parametersRef` was intended/suggested to be used (as a [template](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.GatewayClass)), and the role that it's being stretched into filling in practice because of the gap that adding this Gateway-scoped option for implementation-specific dynamic configuration would help solve.

A summary of the feedback that blocked adding this during API review previously can be found at https://github.com/kubernetes-sigs/gateway-api/pull/1757#issuecomment-1666238150 and additional supporting context for how Gateways are being used (and who is configuring them) to justify this need can be found in https://github.com/kubernetes-sigs/gateway-api/pull/1653#issuecomment-1429992160.

@howardjohn Would this solve the core need in https://github.com/kubernetes-sigs/gateway-api/issues/567?

**Comment by howardjohn** [comment]
@mikemorris from my POV this is good enough to close #567

**Comment by youngnick** [comment]
I'm still a little worried that this will drop the incentive for people to bring further changes to `infrastructure` upstream (since you can now do whatever you want in your own `paramsRef`), but not enough to hold off on this. I would really like to see us consider what things end up in custom `paramsRef` objects that should be lifted out into standard `infrastructure` fields instead.

That said,

/approve

**Comment by howardjohn** [comment]
fwiw I don't think we have plans to implement paramsref in istio in the near term so we are incentivezed to keep pushing for first class fields once this lands :-)

**Review by shaneutt — APPROVED** [review]
/approve
/lgtm
/unhold

**Pull request merged**
